### ISSUE TITLE
ORTE DVM inconsistent behavior

### ISSUE THREAD

**Issue by rhc54** [open-mpi/ompi#4207]
This is an issue on master, and it isn't clear if the issue resides in ORTE or PMIx. If we start the ORTE DVM, the first invocation of prun with any application will execute correctly. However, the next invocation struggles to make a connection to the PMIx server in the ORTE daemon, and then the execution of that application doesn't happen.

Here is an example sequence:
```
$ orte-dvm &
[1] 211830
DVM ready

09:25:19  (topic/prun) /home/common/openmpi/foobar$ prun -n 1 hostname
[rhc001:211894] JOB 2326593538 EXECUTING
rhc001
09:25:22  (topic/prun) /home/common/openmpi/foobar$ prun -n 1 hostname
rhc001
[rhc001:211922] JOB 2326593540 EXECUTING
[warn] Epoll MOD(5) on fd 0 failed.  Old events were 2; read change was 0 (none); write change was 1 (add): Bad file descriptor
09:25:28  (topic/prun) /home/common/openmpi/foobar$ prun -n 1 hostname
<hang>
^C
09:26:35  (topic/prun) /home/common/openmpi/foobar$ prun --terminate
TERMINATING DVM...DONE
[1]+  Done                    orte-dvm
09:26:41  (topic/prun) /home/common/openmpi/foobar$
```
Note that I was seeing "resource busy" when the prun tool attempted to connect to the server, and so I modified the PMIx TCP connection code. When it gets EAGAIN during the blocking recv, it closes the socket, opens another socket, and retries the connection. This is usually adequate to get through, but it is a new disturbing behavior.

I did find one memory corruption in ORTE (due to prun not providing all the necessary app_context fields), and I fixed that one. I believe the file descriptor warning may be due to prun not waiting around until the application actually completes execution - it currently departs as soon as the spawn command completes.

I've spent a bunch of time trying to understand the failure to connect, and have hit a brick wall. At this point, I think additional eyes on the problem would really help. My next plan of attack is to do many cycles of a simple app that calls PMIx_tool_init followed by PMIx_tool_finalize, just to exercise solely the connect/disconnect code and isolate that from the application-execution code.

@jjhursey @dsolt @ggouaillardet @gvallee  If any of you can spare a few cycles to poke at this, I'd appreciate it. The key code path is down in opal/mca/pmix/pmix2x/pmix/src/mca/ptl (both the base and tcp component), though the problem may be due to memory corruption elsewhere.

**Comment by ggouaillardet** [comment]
```prun``` is from the PMIx reference server, right ?
should we use its ```master``` branch ? or one of your branch ?

**Comment by ggouaillardet** [comment]
if i read between the lines, it should be https://github.com/rhc54/ompi/tree/topic/prun

i will try that from now

**Comment by ggouaillardet** [comment]
@rhc54 this is likely unrelated. i am unable to run ```prun``` unless i configure with ```--disable-dlopen``` (some components cannot be loaded because of unresolved symbols)

but i can reproduce the issue when configure'd with ```--disable-dlopen```, and here is a patch that avoids the issue
```patch
diff --git a/opal/mca/pmix/pmix2x/pmix/src/mca/gds/ds12/gds_dstore.c b/opal/mca/pmix/pmix2x/pmix/src/mca/gds/ds12/gds_dstore.c
index 62ce4b1..e781fbe 100644
--- a/opal/mca/pmix/pmix2x/pmix/src/mca/gds/ds12/gds_dstore.c
+++ b/opal/mca/pmix/pmix2x/pmix/src/mca/gds/ds12/gds_dstore.c
@@ -910,7 +910,6 @@ static inline void _esh_session_release(session_t *s)
     }
 
     _delete_sm_desc(s->sm_seg_first);
-    close(s->lockfd);
 
     if (NULL != s->lockfile) {
         if(PMIX_PROC_IS_SERVER(pmix_globals.mypeer)) {
```

when ```_esh_session_release()``` is invoked, ```s->lockfd``` is ```0```, it gets ```close()```d and then bad things happen later. i might have overlooked it (a bunch of macros are involved) but so far, it seems ```lockfd``` is not set anywhere in the code (e.g. the ```gds/ds12``` component)

**Comment by rhc54** [comment]
I'll fix the link problem (missing a link in the Makefile.am). I could swear we had fixed that darn close once before, yes?? Maybe it didn't get upstreamed properly, or it got overwritten. Sigh.

Thanks for tracking it down!

**Comment by rhc54** [comment]
I see what happened - we fixed it in PMIx v2.0, and it somehow got overwritten/lost in PMIx master.

**Comment by rhc54** [comment]
Fixed by #4217
